### PR TITLE
Add core typing to documentation

### PR DESCRIPTION
**Pull request by user27182**
### Overview

Add documentation for the type definitions in `core._typing_core`. 

Specifically:

1. Publish the types in the docs in `Core API -> Typing`
2. Make the `_typing_core` types in docstrings and annotations automatically linkable.
    - Example issue: the docs for [set_array](https://docs.pyvista.org/version/0.43/api/core/_autosummary/pyvista.DataSetAttributes.set_array.html) have a `data : Array` parameter in the docstring, but it is not linkable. This makes it unclear what is meant by `Array`.
    - This is accomplished by adding core types to the `pyvista` namespace. This is so that adding the types to the docs with just the type name (e.g. `Vector` will mean that sphinx will implictly find and link it automatically (no need to explicitly reference `pyvista.core._typing_core.Vector`)

3. Make the `_typing_core` annotations in function signatures more readable.
    - Example issue: since aliases are automatically expanded by sphinx, the annotations for array-like types can take up to 8 lines for a single parameter. See screenshot in https://github.com/pyvista/pyvista/pull/5571#issuecomment-1939197128

The additions in this PR are similar to the configuration for [ColorLike](https://docs.pyvista.org/version/stable/api/utilities/_autosummary/pyvista.ColorLike.html), which is a pyvista type alias that is linkable and is defined formally the docs. (See the `color` param for [ChartBox](https://docs.pyvista.org/version/stable/api/plotting/charts/_autosummary/pyvista.chartbox#pyvista.ChartBox) as an example of its use in docstrings).

### CODE REVIEW

**Review by tkoyama010 — DISMISSED** [review]
LGTM

**Review by tkoyama010 — APPROVED** [review]

File: pyvista/core/_typing_core/__init__.py [hunk 1]
```diff
@@ -10,4 +10,35 @@
...
+    "Includes sequences, nested sequences, and numpy arrays up to four dimensions.\n"
+    "Scalar values are not included."
+)
+TransformLike.__doc__ = "NumPy array or vtk object representing a 3x3 or 4x4 matrix."
+BoundsLike.__doc__ = (
+    "Tuple of six values representing 3D bounds.\n"
+    "Has the form (xmin, xmax, ymin, ymax, zmin, zmax)."
+)
+# CellsLike.__doc__
+# CellArrayLike.__doc__
```
Comment (user27182) [inline pyvista/core/_typing_core/__init__.py:44]: I wasn't sure how to document these. @darikg do you have a suggestion for these?

**Comment by user27182** [comment]
Thanks for the fixes and quick approval!

Might be good to do a manual check before this is merged. The aliases need `from __future__ import annotations` to be documented properly by sphinx. But this is only defined in `_alias`, whereas the `__doc__`s  are added in `__init__`.

So, I'm not sure if sphinx will actually pick up the docstrings correctly or not. I'm also curious to see how the `TypeVar` will appear in the docs.

**Review by tkoyama010 — CHANGES_REQUESTED** [review]
Thanks. I will check in my local. It would be nice if we could do a preview build, but that mechanism is being rebuilt in #5360.

**Comment by user27182** [comment]
I have not been able to review the docs locally. The docs are approx 3.5GB in size, and download speeds seem to be capped at 500KB/sec, which means downloading them takes approx 2 hours. For some reason, the download keeps failing for me part way through and I have to re-retry. The download has failed 4 times for me already, not sure if that's just my machine or my web browser.

**Comment by tkoyama010** [comment]
> I have not been able to review the docs locally. The docs are approx 3.5GB in size, and download speeds seem to be capped at 500KB/sec, which means downloading them takes approx 2 hours. For some reason, the download keeps failing for me part way through and I have to re-retry. The download has failed 4 times for me already, not sure if that's just my machine or my web browser.

The volume of documentation is bloated and is straining our development efficiency. This is a significant problem. We are discussing it in #5352 and would appreciate any ideas you may have.

**Comment by user27182** [comment]
I still can't download the docs locally for review 😔, not sure what the issue is or if there are any workarounds...

**Comment by tkoyama010** [comment]
github-actions preview

**Comment by tkoyama010** [comment]
github-actions preview

**Comment by tkoyama010** [comment]
> 🚀 Deployed on https://65d1e17a91a5a723b3ec9543--meek-duckanoo-d2f5ec.netlify.app

Preview is deployed. Let's review!

**Comment by tkoyama010** [comment]
github-actions preview

**Comment by user27182** [comment]
> 🚀 Deployed on https://65d306c928293de4dfcc21a8--meek-duckanoo-d2f5ec.netlify.app

aliases are still expanded and not linked (I am checking the docs for `pyvista.DataSetAttributes.set_array` specifically), and __doc__ strings aren't shown either. Maybe https://github.com/pyvista/pyvista/pull/5621/commits/bb86bf6553a9466cd573230250f0f0077421ba58 will fix it?

**Comment by tkoyama010** [comment]
github-actions preview

**Comment by user27182** [comment]
The docs still have the same issues: aliases are still expanded, and docstrings aren't showing.

The PR which added ColorLike (PR #2175) also added exclusions to `numpydoc_validation_exclude`... maybe that's the issue, and new exclusions need to be added ??

**Comment by tkoyama010** [comment]
github-actions preview

**Comment by tkoyama010** [comment]
Feel free to comment on the commands to see preview. Your account also has the authority I guess.

**Comment by user27182** [comment]
github-actions preview

**Comment by user27182** [comment]
github-actions preview

**Comment by user27182** [comment]
github-actions preview

**Comment by user27182** [comment]
github-actions preview

**Comment by user27182** [comment]
github-actions preview

**Comment by user27182** [comment]
Finally found the magic combination for the docs. Type alias docstrings are showing, and the `Array` object in `datasetattributes.set_array` now links directly to the Alias. 

LGTM

**Comment by user27182** [comment]
github-actions preview

**Comment by user27182** [comment]
> Deployed on https://65d830adf48efa981506a497--meek-duckanoo-d2f5ec.netlify.app

It looks like the links are no longer working with `datasetattributes`. I suspect this may be from 5dbc9bc, which removed the `from __future__ import annotations` statement. This could mean that it may be necessary to add that line to _every_ file where the alias is used.

**Comment by user27182** [comment]
github-actions preview

**Comment by user27182** [comment]
> 🚀 Deployed on https://65d9097e1d98156ae7d89bd8--meek-duckanoo-d2f5ec.netlify.app

The alias in the set_array function signature is now correctly set as `ArrayLike`. 

The alias in signature is not linked, however. This seems to be a bug, with a possible workaround here.
https://github.com/sphinx-doc/sphinx/issues/10785

**Comment by user27182** [comment]
github-actions preview

**Comment by user27182** [comment]
This PR looks good to me. 

I am a bit concerned that some references to `ArrayLike` might not get picked up correctly in the future for modules that will import `ArrayLike`. But, this issue likely applies to `ColorLike` as well. So,I think maybe a separate PR can work to resolve that if or when that happens.

**Review by tkoyama010 — APPROVED** [review]
LGTM. Thanks!

**Comment by tkoyama010** [comment]
Please wait while we merge #5670 first.

**Comment by user27182** [comment]
> I am a bit concerned that some references to `ArrayLike` might not get picked up correctly in the future for modules that will import `ArrayLike`. But, this issue likely applies to `ColorLike` as well. So,I think maybe a separate PR can work to resolve that if or when that happens.

Unfortunately, this issue has surfaced. Looking at the docs published here before this PR was merged,
https://65d951a0b791cd8a5fda3017--meek-duckanoo-d2f5ec.netlify.app/api/core/_autosummary/pyvista.polydata

we see that PolyData's array-like aliases are expanded.

image

**Comment by user27182** [comment]
I wonder if @dcbr can help.

This PR added `ArrayLike` aliases to the docs, similar to what was done for `ColorLike` in #2175.
But, we are having issues where it seems that `from __future__ import annotations` may need to be added to every file where we want the alias to display correctly in the docs. Ideally we wouldn't need to do this.

If we look at the docs for [ChartBox](https://docs.pyvista.org/version/stable/api/plotting/charts/_autosummary/pyvista.chartbox#pyvista.ChartBox), it has a working `ColorLike` alias in the parameters. And, looking at the source code, I see that `charts.py` does _not_ have a the `from future` import statement.

Do you know how this was achieved?
Not sure if there's a trick with how the alases are imported, or a config hack that was missed by this PR (but is part of #2175).

**Comment by dcbr** [comment]
It's been a while since I messed with those aliases, I know it was a lot of trial and error and don't remember the exact "solution"; but I will try to have another look and see if I can help resolve it.

**Comment by user27182** [comment]
> It's been a while since I messed with those aliases, I know it was a lot of trial and error and don't remember the exact "solution"; but I will try to have another look and see if I can help resolve it.

Thanks, this PR tried to copy the same config, and there was also lots of trial and error. A possible source of the issue may be from:
1. differences in how the aliases are defined and imported into the `pyvista` namespace
    - `ColorLike` gets imported lazily (like many `plotting` objects`), whereas `ArrayLike` gets imported explicitly via `core._typing_core.__init__`
    - This may be relevant since `from __future__ import annotations` is definitely needed in the modules where the aliases are defined, so maybe the import messes with this somehow
2. the `plotting._typing` modules imports these explicitly while `TYPE_CHECKING`: https://github.com/pyvista/pyvista/blob/e338c17ba85b7bc2c9cd4903c9a3ebd855555c83/pyvista/plotting/_typing.py#L11-L13
    - These imported modules use `ColorLike`, so perhaps that's what allows the docs to recognize the aliases
    - The `ArrayLike` modules do not have this kind of import 


After writing this, I strongly suspect that # 2 is the issue, and that adding a conditional `TYPE_CHECKING` import for any/all modules which may use the `ArrayLike`  aliases may fix this.

File: doc/source/api/core/typing.rst [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+Type aliases and type variable used by PyVista.
+
+
+Numeric Array-Like Types
+------------------------
+
+.. autotypevar:: NumberType
+
+.. autodata:: ArrayLike
```
Comment (dcbr) [inline doc/source/api/core/typing.rst:13]: Sorry, I haven't had much time lately, so I haven't properly looked into this yet.
But another difference I noticed is the way the documentation for the type alias is generated here. For `ColorLike` it is generated using `.. autosummary::` (see utilities.rst). I'm not at all a Sphinx expert, so maybe these are equivalent, but maybe you can give this a try?
Additionally, I don't see this typing module being documented on [the deployed documentation for the 'dev' branch](https://docs.pyvista.org/version/dev/api/core/index.html)?

**Pull request merged**